celery/app/task.py

```diff
@@ -769,11 +774,14 @@ def apply(self, args=None, kwargs=None,
             task.name, task, eager=True,
             propagate=throw, app=self._get_app(),
         )
+
         ret = tracer(task_id, args, kwargs, request)
         retval = ret.retval
+        state = states.SUCCESS if ret.info is None else ret.info.state
         if isinstance(retval, ExceptionInfo):
             retval, tb = retval.exception, retval.traceback
-        state = states.SUCCESS if ret.info is None else ret.info.state
+        if isinstance(retval, RetryEagerException):
+            return retval.signature.apply()
```

Comment:
So essentially here we detect that a `Retry` exception type has been raised and call the task signature synchronously again?